lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isBreakStatement(Statement statement) => (statement is BreakStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is BreakStatement));
+
+bool _isContinueStatement(
+        Statement statement) =>
+    (statement is ContinueStatement ||
+        (statement is Block &&
+            statement.statements.length == 1 &&
+            statement.statements.first is ContinueStatement));
+
+bool _isReturnStatement(Statement statement) => (statement is ReturnStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is ReturnStatement));
+
+abstract class ConditionScopeVisitor extends RecursiveAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
```

Comment:
In this case, we need an assignment expression whose left-hand-side is not resolved:
```
notDefined = 3;
```
(any name that isn't defined will work). Of course, you'll want to back out your fix to ensure that the test case actually triggers the bug.